### PR TITLE
update

### PR DESCRIPTION
**Pull request by sdz7121211**
1. add Merge engine
2. add ReplacingMergeTree
3. Implements a client

### CODE REVIEW

**Comment by ishirav** [comment]
Hi,

Thank you for contributing to this project, but this pull request is not in an acceptable state. If you want to add support for Merge and ReplacingMergeTree, please add only the relevant files. Currently there are irrelevant files (under .idea), a ClickHouseClient that seems to be used only for testing yet is not under /tests, changes to datetime fields that use a specific timezone, etc.

Thanks

**Comment by sdz7121211** [comment]
add insert_try_best func, ensure bulk insert data as far as possible, ignore error entries.
add ReplacingMergeTree and Merge engine.

File: src/infi/clickhouse_orm/database.py [hunk 1]
```diff
@@ -12,8 +12,10 @@
 logger = logging.getLogger('clickhouse_orm')
 
 
+
 Page = namedtuple('Page', 'objects number_of_objects pages_total number page_size')
 
+logger = logging.getLogger(__name__)
```
Comment (ishirav) [inline src/infi/clickhouse_orm/database.py]: logger is already defined above

File: src/infi/clickhouse_orm/database.py [hunk 2]
```diff
@@ -79,6 +81,56 @@ def gen():
                 yield ('\n'.join(batch) + '\n').encode('utf-8')
         self._send(gen())
 
+    def insert_try_best(self, model_instances, batch_size = 1000, is_create = False, insert_count = 0, first_recu = True):
```
Comment (ishirav) [inline src/infi/clickhouse_orm/database.py]: What is this method, why is it needed, what kind of problems does it try to solve?

File: src/infi/clickhouse_orm/database.py [hunk 3]
```diff
@@ -79,6 +81,56 @@ def gen():
...
+                insert_count += len(cache)
+            except:
+                # binary insertion
+                if len(cache) > 1:
+                    self.insert(self, cache[:(len(cache) / 2)], min(len(cache[:(len(cache) / 2)]) / 2, batch_size),
+                                is_create=is_create, insert_count=insert_count, first_recu=False)
+                    self.insert(self, cache[(len(cache) / 2):], min(len(cache[(len(cache) / 2):]) / 2, batch_size),
+                                is_create=is_create, insert_count=insert_count, first_recu=False)
+                elif len(cache) == 1:
+                    logger.error("%s; LOG: %s" % (str(sys.exc_info()), ))
```
Comment (ishirav) [inline src/infi/clickhouse_orm/database.py]: You can use logger.exception which already prints out the traceback on its own

File: src/infi/clickhouse_orm/fields.py [hunk 4]
```diff
@@ -119,7 +119,7 @@ def to_db_string(self, value, quote=True):
 
 class DateTimeField(Field):
 
-    class_default = datetime.datetime.fromtimestamp(0, pytz.utc)
+    class_default = datetime.datetime.fromtimestamp(0, pytz.timezone('Asia/Shanghai'))
```
Comment (ishirav) [inline src/infi/clickhouse_orm/fields.py]: No, you cannot change the timezone

File: ClickHouseClient/ClickHouseClient.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
```
Comment (ishirav) [inline ClickHouseClient/ClickHouseClient.py]: What is the purpose of the ClickHouseClient and all the rest of the files in this folder? Is this some sort of test?

File: src/infi/clickhouse_orm/database.py [hunk 6]
```diff
@@ -79,6 +81,56 @@ def gen():
...
+            if len(cache) >= batch_size:
+                try:
+                    # Must ensure that batch_size > = bulksize
+                    self.insert(cache, batch_size=batch_size)
+                    insert_count += len(cache)
+                    cache = []
+                except:
+                    # binary insertion
+                    if len(cache) > 1:
+                        self.insert(self, cache[:(len(cache) / 2)],
```
Comment (ishirav) [inline src/infi/clickhouse_orm/database.py]: Calling self.insert won't work, it doesn't have parameters called bulksize, is_create, etc. Perhaps you meant to call insert_try_best recursively?

File: src/infi/clickhouse_orm/fields.py (hunk 4 above)
Comment (emakarov) [inline src/infi/clickhouse_orm/fields.py]: please switch back the default timezone to UTC

File: ClickHouseClient/ClickHouseClient.py (hunk 5 above)
Comment (emakarov) [inline ClickHouseClient/ClickHouseClient.py]: this file seems like your local file, please do not add it to the repository

File: ClickHouseClient/conf.ini [hunk 7]
```diff
@@ -0,0 +1,3 @@
+[clickhouse]
```
Comment (emakarov) [inline ClickHouseClient/conf.ini]: this seems to be your local files. please remove them from merge request and leave only required functionality

File: src/infi/clickhouse_orm/database.py [hunk 8]
```diff
@@ -79,6 +81,56 @@ def gen():
...
+                is_create = True
+                table_name = elem.table_name()
+            cache.append(elem)
+            if len(cache) >= batch_size:
+                try:
+                    # Must ensure that batch_size > = bulksize
+                    self.insert(cache, batch_size=batch_size)
+                    insert_count += len(cache)
+                    cache = []
+                except:
```
Comment (emakarov) [inline src/infi/clickhouse_orm/database.py]: please add expected exception type to all except blocks

File: src/infi/clickhouse_orm/database.py (hunk 2 above)
Comment (emakarov) [inline src/infi/clickhouse_orm/database.py]: please add docstring to the function

File: src/infi/clickhouse_orm/database.py (hunk 6 above)
Comment (sdz7121211) [inline src/infi/clickhouse_orm/database.py]: my mistake, as you said should be call insert_try_best recursively, 
this func to ensure insert when a bulk insert faild because of some dirty data.

File: src/infi/clickhouse_orm/fields.py [hunk 9]
```diff
@@ -128,6 +128,7 @@ def to_python(self, value, timezone_in_use):
         if isinstance(value, datetime.date):
             return datetime.datetime(value.year, value.month, value.day, tzinfo=pytz.utc)
         if isinstance(value, int):
+            # return datetime.datetime.fromtimestamp(value, pytz.timezone('Asia/Shanghai'))
```
Comment (emakarov) [inline src/infi/clickhouse_orm/fields.py]: this commented line should be removed

File: src/infi/clickhouse_orm/models.py [hunk 10]
```diff
@@ -48,7 +48,7 @@ def create_ad_hoc_model(cls, fields):
 
     @classmethod
     def create_ad_hoc_field(cls, db_type):
-        import infi.clickhouse_orm.fields as orm_fields
+        import fields as orm_fields
```
Comment (emakarov) [inline src/infi/clickhouse_orm/models.py]: why this change of import is done?

File: src/infi/clickhouse_orm/database.py [hunk 11]
```diff
@@ -79,6 +82,56 @@ def gen():
                 yield ('\n'.join(batch) + '\n').encode('utf-8')
         self._send(gen())
 
+    def insert_try_best(self, model_instances, batch_size = 1000, is_create = False, insert_count = 0, first_recu = True):
+        cache = []
+        table_name = None
+        for elem in model_instances:
+            if is_create == False:
```
Comment (emakarov) [inline src/infi/clickhouse_orm/database.py]: here you're using variable is_create and then make `create_table` if `is_create == False`. This seems to be non-obvious behaviour. `is_create` sounds more like something should be created, but you're doing it vise-versa. Please choose another name for variable or change logic. In general, purpose of this recursive method is not clear for me now. Please describe this code.

**Comment by ishirav** [comment]
@sdz7121211 - you will need to communicate and explain the purpose behind all this code you've committed. It is not clear what it's for, and to me it looks like it doesn't belong in this project. Perhaps you should create a separate project for it, which uses infi.clickhouse_orm as a dependency and builds additional functionality in top of it.
